Add support for custom player models

cjohnson57 opened the pull request:
Adds support for selecting (on local) or uploading via filepicker (on web) a .zobj definition for a custom player model. Unlike ML64, these models are subject to size and poly count restrictions present in the normal game. There are two types of models, which will be referred to as zzconvert and zzplayas models.

### zzconvert models 
are models which are output from Blender with a custom extension. They only contain original data created by the modeler. Thus, they can be freely shared through the Discord or other means without worries of copyright. In zzconvert models, the LUT is empty. (LUT = LookUp Table, a table in the model that points to the displaylists for all pieces of it.)

### zzplayas models
are one step further in the process, having gone through a process to add all pieces missing from the zzconvert model into it, which is copyrighted content. For example, if the modeler did not create a custom hookshot, zzplayas will add the hookshot from a bank file into the model.

Since zzplayas uses a bank file which is basically just the vanilla Link model, and thus copyrighted content, the code here recreates the functionality of zzplayas to some extent, but it pulls the vanilla data from the user's rom so there should be no copyright concerns.

On local, models can be placed in data/Models/Adult or data/Models/Child and a combobox will fill with the .zobj files in those folders, allowing you to select whichever model you want to use, or just choosing Default for the default player model. If you have more than 1 custom model, there is also a Random option to randomly choose among your custom models. On web, there are instead filepickers where you select the path to your model file to be loaded in. @TreZc0 from my understanding, you and/or db0 will need to add some support for the filepicker to the website once this is merged.

Once the model is loaded from the file, it first checks if it's a playas model (by looking for the string "MODLOADER64" at the beginning of the LUT). If it is, then it needs no further processing and is slapped into the player model portion of the ROM and all the bits are repointed. If it isn't, then the model's footer (created by zzconvert) is checked to see which pieces are present in the model and which are missing and need to be loaded from vanilla. After loading these pieces, the LUT is created by writing pointers to individual pieces as well as writing some constants. It also creates the hierarchy pointer, which is what controls how the different pieces are loaded in the game and relate to each other.

When loading the vanilla data, the displaylist is not enough. It will also contain pointers to texture, matrix, and vertex data, so those will need to be loaded as well, and then repointed in the displaylist so it knows where to look for this data. The vanilla data basically creates a 2nd zobj with just the missing pieces, which is then slapped onto the end of the user-provided zobj and then finally loaded into the ROM.

Certain indices of some pieces are skipped to remove redundancy or unnecessary vanilla data. For example, most vanilla items have Link's fist attached to them, but as zzconvert's hierarchy simply uses the default empty fist and adds the item into it, this is unnecessary- All it does is stick Link's fist inside of whatever the given model's fist is. So the bytes of an item's displaylist that contain the fist are skipped so that only the item remains.

Note that the code here doesn't change the skeleton at all, meaning hit and hurtboxes will all be where they are normally, so a model should offer no competitive advantage (other than if someone makes a model that gives them better visual cues or something I guess.)

This has been tested with a few different zzconvert models so far, as well as on Everdrive.

I tried to comment everything very in-depth since this code is complex and not really part of anyone's existing skillset at OoTR, but feel free to ask here or @ me in the discord if there's something confusing or just needing more explanation.

-----------

Ginormous thanks to @Maplesstar for making models to use, helping me figure out issues, and general emotional support. Also since she will be the one making the guides for model creation so that this feature is easier to use.

Gigantic thanks to @denoflionsx for helping point me in the right direction, understand the ML64 model code, helping sleuth issues, and more- Same goes for @JerryWester, @Neirn, and other helpful folks over at ML64.

Note: These screenshots were all captured from Project64, and these are all zzconvert models.

![GLideN64_THE_LEGEND_OF_ZELDA_002](https://user-images.githubusercontent.com/30701749/162876298-9208b367-d95d-4a0b-b100-05a1e5d6da3e.png)

![GLideN64_THE_LEGEND_OF_ZELDA_010](https://user-images.githubusercontent.com/30701749/162876396-4a985f8a-44ed-4e36-9536-6878ac2c98f3.png)

![GLideN64_THE_LEGEND_OF_ZELDA_005](https://user-images.githubusercontent.com/30701749/162876212-84ad2a48-dd02-4226-8bb8-adf59baa9ecb.png)

![GLideN64_THE_LEGEND_OF_ZELDA_008](https://user-images.githubusercontent.com/30701749/162876326-642d5eb8-a444-4041-bf7a-006c22134c86.png)

![GLideN64_THE_LEGEND_OF_ZELDA_007](https://user-images.githubusercontent.com/30701749/162876219-5bc252ef-5055-48cc-929c-e887042b0c28.png)

cjohnson57 committed:
Add models selection which pulls from data/Models, also a random option

cjohnson57 committed:
Recreate functionality of manifest code

Currently crashes when loading title screen

cjohnson57 committed:
Use shields in place of DFs, write child hierarchy

cjohnson57 committed:
Fix missing rfist and write float hex directly

denoflionsx committed:
Fix hierarchy and pointers

cjohnson57 committed:
Merge pull request #1 from denoflionsx/models

Fix hierarchy and pointers

cjohnson57 committed:
Separate child and adult model functionality

cjohnson57 committed:
Fix missing slingshot and adult gauntlet fist; add child and adult model pieces

cjohnson57 committed:
Experimental code to pull vanilla model data

Need more work for pointers and stuff

cjohnson57 committed:
Optimize: DL bytecode crawling

cjohnson57 committed:
Add rest of optimize code; don't add vanilla data until post-optimization

cjohnson57 committed:
Add LUT and vanilla data to pieces

cjohnson57 committed:
Run optimize on Link's full vanilla zobj

cjohnson57 committed:
Set LUT for all pieces

cjohnson57 committed:
Set hierarchy pointer

cjohnson57 committed:
Write constant data to zobj; hierarchy pointer unwraps once

cjohnson57 committed:
Don't process PlayAs Models

cjohnson57 committed:
Write DF in Optimize, add LUT prefix, change verbiage

cjohnson57 committed:
Use proper rebase value

Models mostly work now! Just have to get rid of extra parts

cjohnson57 committed:
Remove adult duplicates by specifying skipped bytes

cjohnson57 committed:
Properly write DFs where they belong

cjohnson57 committed:
Remove duplicated parts from child and clean up adult

cjohnson57 committed:
Zero out LUT space before processing models

Fixes child models crashing

cjohnson57 committed:
Update GUI

Simplify model choice loading code, put model choices in their own GUI pane

cjohnson57 committed:
Clean up child

cjohnson57 committed:
Merge remote-tracking branch 'upstream/Dev' into models

cjohnson57 committed:
Fix botched merge

cjohnson57 committed:
Check for Hookshot and Blade.3 in scan()

cjohnson57 committed:
Fix fists in scan, make "Hookshot" the far model

cjohnson57 committed:
Clean up: Errors, documentation, renaming, style

cjohnson57 committed:
Add filepickers for web

cjohnson57 committed:
Clean up and comments

cjohnson57 committed:
Make sure only .zobjs are being grabbed

denoflionsx commented:
I ensured while guiding Shadow through the initial stages of this process that all existing ML64 models that meet the N64 specifications will work. Users of OotR are free to use our [repository of models](https://github.com/hylian-modding/Z64-CustomPlayerModels) to obtain any model available to ML users in raw zobj form.

flagrama commented:
Has this been tested at all on Virtual Console and Everdrive yet?

Maplesstar commented:
No, as of right now tested on PJ64 and Retroarch. I intend to do VC tests soon if no one else does first, but everdrive we need someone else.

flagrama commented:
I did some quick testing with Everdrive. Nothing extensive, but confirmed nothing crashed immediately and went around various areas as child and adult custom models. Did run into a graphical glitch where the Master Sword disappeared when pulling it from the pedestal, but the game didn't crash so seems like it's just a cosmetic issue. Aside from that, someone else should probably do an in-depth test with Everdrive, but it at least runs and so far doesn't seem different than playing with the vanilla model.

cjohnson57 commented:
Confirmed the MS-pulling issue isn't everdrive exclusive, will probably have a fix up tomorrow

malonrose commented:
Amazing work everyone!
Super excited to see this implemented 👍

cjohnson57 committed:
Fix pulling the MS

cjohnson57 committed:
Check if skeleton is vanilla

If close to vanilla, forcibly set it to vanilla

cjohnson57 committed:
Draw illegal model text on pause screen if skeleton bad

cjohnson57 commented:
The code will now check the model's skeleton. If it's very close to the vanilla skeleton, it'll simply be corrected to be vanilla. If not, it draws a message in the pause screen that the model does not have a vanilla skeleton and is thus illegal for racing.

fenhl commented on ASM/c/model_text.c:
```diff
@@ -0,0 +1,33 @@
+    
+    // Setup draw location
+    int str_len = 36;
+    int total_w = str_len * font_sprite.tile_w + triforce_sprite.tile_w;
```
```suggestion
    int total_w = str_len * font_sprite.tile_w;
```

cjohnson57 replied:
Oops, you can tell where I copied this code from, lol

cjohnson57 committed:
Oh no I've revealed my plagiarism

cjohnson57 committed:
Change message wording

cjohnson57 commented:
Here's what the non-vanilla skeleton message looks like in game:

![image](https://user-images.githubusercontent.com/30701749/164585483-13b38baa-d4d8-4897-ac9d-5e9f06060e32.png)

A bit close to the margins unfortunately, but since I can't control font size and don't want it covering any part of the menu or the triforce display, I think this is the best we can do.

cjohnson57 committed:
Ensure model and triforce text can both be drawn

Maplesstar committed:
Add Tooltip for Model Selection

cjohnson57 committed:
Change message wording

Don't specify "illegal" in case allowed in some races

cjohnson57 committed:
Merge pull request #2 from Maplesstar/models, ty maple

Add Tooltip for Model Selection

cjohnson57 committed:
make text white

cjohnson57 committed:
Merge branch 'models' of https://github.com/cjohnson57/OoT-Randomizer into models

cjohnson57 committed:
Update NPC masks

cjohnson57 committed:
Patch models anytime cosmetics are patched

Fixes override cosmetics not patching models in

cjohnson57 committed:
Fix hiding model things when generating from patch file

cjohnson57 committed:
Make patching a model over a patch file with model work

Before applying patch file to rom save vanilla model and reapply it after so the model code that deals with vanilla model data will work

cjohnson57 commented:
Using patch files should now work, except for the case where a patch file contains a model and you attempt to override it with a vanilla model. I can't think of any good way to do this currently as all the pointers are overridden to use their ml64 manifest (LUT) values in the patch file, and I'd just need to somehow set them all back to their vanilla values.

cjohnson57 committed:
Ignore certain model-related bytes when applying patch file and overriding cosmetics

cjohnson57 commented:
Ok, all issues with patch files should now be fixed.

cjohnson57 committed:
Merge remote-tracking branch 'upstream/Dev' into models

cjohnson57 committed:
Recompile c after merge

Maplesstar committed:
Add instructions in ReadMe

cjohnson57 committed:
Merge pull request #3 from Maplesstar/models

Add instructions to the ReadMe

cjohnson57 committed:
Fix gauntlet piece offsets

Maplesstar committed:
ReadMe note about incompatible models

cjohnson57 committed:
Fix hilt offset

cjohnson57 committed:
Merge remote-tracking branch 'upstream/Dev' into models

cjohnson57 committed:
Merge pull request #4 from Maplesstar/models

ReadMe note about incompatible models

cjohnson57 committed:
quick web fixes

cjohnson57 committed:
Merge branch 'models' of https://github.com/cjohnson57/OoT-Randomizer into models

mracsys committed:
append file ext to filename with random setting

cjohnson57 committed:
Merge pull request #5 from mracsys/random_models_fix

Fix Random option for model selection

cjohnson57 committed:
Fix swapped hookshot models

cjohnson57 committed:
Merge branch 'Dev' into models

cjohnson57 committed:
Fix crash on equipment screen when child has hylian shield but no sword

cjohnson57 committed:
Merge remote-tracking branch 'upstream/Dev' into models

cjohnson57 committed:
Merge remote-tracking branch 'upstream/Dev' into models

cjohnson57 committed:
Add segment to DF address so it works

cjohnson57 committed:
Merge remote-tracking branch 'upstream/Dev' into models

cjohnson57 committed:
Merge remote-tracking branch 'upstream/Dev' into models

cjohnson57 committed:
Hide models from web gen page

Maplesstar committed:
Alternate Voices for Link

Add SFX options to independently change adult and child voices between default, silent, and feminine. Currently won't let you change both voices off default at once.

Maplesstar committed:
All Voice Combos Now Work

Required going to a single .zpf for each combo, but it's fiiiiine.

Maplesstar committed:
Add patch file issue with voices to readme

Maplesstar committed:
heckin typo in the readme

Maplesstar committed:
Fix Patch File Issues

Combined the voice function into the overall cosmetics function and reset audiobank/table bytes to the base rom before patching voices. (Also corrects a voice issue with the adult female voice .zpfs.)

Maplesstar committed:
Merge branch 'models' of https://github.com/Maplesstar/OoT-Randomizer into models

Maplesstar committed:
Update README.md

Maplesstar committed:
Separate override voice and correct Voice .zpfs. Again.

Made a separate function for the voice override after a patch file so that going back to default/default actually happens despite no voice patches applied. Also ended up re-doing all the .zpfs just to double triple check they're right.

Maplesstar committed:
Merge branch 'models' of https://github.com/Maplesstar/OoT-Randomizer into models

cjohnson57 committed:
Remove setting flag for illegal model; add some mostly commented code for checking model in rom

cjohnson57 committed:
Merge pull request #8 from Maplesstar/models

Voice Replacement

cjohnson57 committed:
Merge remote-tracking branch 'upstream/Dev' into models

cjohnson57 committed:
Voice default values

Cuphat committed:
Use data_path for get_model_choices.

Cuphat committed:
Use data_path/os.path.join in more places.

cjohnson57 committed:
Fix up voice patches

Remove overriding early bytes after changing the voice zpfs, remove voices from randomize all sfx, add random choice option for voices, move voices into their own folder in data

cjohnson57 committed:
Merge remote-tracking branch 'upstream/Dev' into models

cjohnson57 commented:
Note that while this PR has been merged, we still need to find a solution for adding the race-illegal skeleton message to the ROM from the webpatcher.
